### PR TITLE
Implemented the df64 floating point arithmetic for high precision layers

### PR DESCRIPTION
**Pull request by shaojingli**
Please review. @gnavvy @ibgreen

This change requires luma.gl change to function properly.

### CODE REVIEW

File: src/layers/fp64/scatterplot-layer/scatterplot-layer.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  df64ify(a) {
+    const a_hi = new Float32Array([a])[0];
+    const a_lo = a - a_hi;
+    return [a_hi, a_lo];
+  }
+
+  updateUniforms() {
+    this.calculateZoomRadiusFP64();
```
Comment (ibgreen) [inline src/layers/fp64/scatterplot-layer/scatterplot-layer.js]: Not sure that we need to change the name of this method. The fact that the zoom radius happens to be stored in 64 bits is a detail inside that function.

File: src/layers/shaderlib/fp64/mul-fp64.glsl [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+vec2 split(float a) {
```
Comment (ibgreen) [inline src/layers/shaderlib/fp64/mul-fp64.glsl]: I see this function in multiple places (also in `div...`) - does that cause problems when importing both? should we have a common file with such utils?

File: src/layers/shaderlib/fp64/vec4-sum-fp64.glsl [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+#pragma glslify: sum_fp64 = require(./sum-fp64)
+
+void vec4_sum_fp64(vec2 a[4], vec2 b[4], out vec2 out_val[4]) {
+
+  for (int i = 0; i < 4; i++)
+  {
```
Comment (ibgreen) [inline src/layers/shaderlib/fp64/vec4-sum-fp64.glsl]: brace on preceding line, remove empty lines/return etc.

**Comment by contra** [comment]
Just wanted to say thank you! Really appreciate the work you did to fix this. This extra layer of polish will go a **long** way. 🎉

File: src/react/deckgl-overlay.js [hunk 4]
```diff
@@ -145,6 +145,27 @@ export default class DeckGLOverlay extends React.Component {
       }
     }
```
Comment (shaojingli) [inline src/react/deckgl-overlay.js]: I agree with this. But we could have it separately later.

File: src/layers/fp64/scatterplot-layer/scatterplot-layer.js [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  initializeState() {
+    const {gl} = this.state;
+    const {attributeManager} = this.state;
+
+    this.setState({
+      model: this.getModel(gl)
+    });
+
+    attributeManager.addInstanced({
+      instancePositions: {size: 4, update: this.calculateInstancePositions},
```
Comment (shaojingli) [inline src/layers/fp64/scatterplot-layer/scatterplot-layer.js]: Changed to instancePositionsFP64

File: src/layers/fp64/scatterplot-layer/scatterplot-layer.js (hunk 1 above)
Comment (shaojingli) [inline src/layers/fp64/scatterplot-layer/scatterplot-layer.js]: Yes. I changed that to have all function names/uniform names following the same pattern.

**Comment by shaojingli** [comment]
I've fixed most of the coding-style issues mentioned in comments inline in latest revision. I'll open another PR when I'm done moving fp64 math functions to a standalone library.

**Comment by shaojingli** [comment]
I've also merged the performance testing layers implemented in dev to fp64. It turns out the fp64 layers can render 1M points in scatterplot with usable framerate.

**Comment by shaojingli** [comment]
I also plan to add shader accuracy tests in next revision. Please review, @ibgreen @gnavvy

File: src/layers/fp64/scatterplot-layer/scatterplot-layer.js [hunk 6]
```diff
@@ -104,19 +105,20 @@ export default class ScatterplotLayer extends BaseLayer {
   }
 
   df64ify(a) {
-    const a_hi = new Float32Array([a])[0];
-    const a_lo = a - a_hi;
+    // const a_hi = new Float32Array([a])[0];
+    const a_hi = Math.fround(a);
```
Comment (gnavvy) [inline src/layers/fp64/scatterplot-layer/scatterplot-layer.js]: +1, was not aware of this function.

File: src/layers/fp64/scatterplot-layer/scatterplot-layer.js [hunk 7]
```diff
@@ -143,6 +146,7 @@ export default class ScatterplotLayer extends BaseLayer {
       [value[i + 0], value[i + 1]] = this.df64ify(position[2]);
       i += size;
     }
+    console.log("height: ", value);
```
Comment (gnavvy) [inline src/layers/fp64/scatterplot-layer/scatterplot-layer.js]: not sure if missed some previous comments, but shall we remove the console.logs here and below?

Comment (shaojingli) [inline src/layers/fp64/scatterplot-layer/scatterplot-layer.js]: oops.... this is my bad...

**Pull request merged**